VM CrashLoop Detection and Exponential Backoff

davidvossel opened the pull request:
# CrashLoop detection

A VM crashloop is being defined as a VM with runStrategy = Always|RerunOnFailure with VMIs that continually fail to reach `vmi.Status.Phase == Running`... Meaning a VM's VMI continually gets scheduled and fails before ever successfully launching the qemu process.

During such an event, the VM crashloop detection will begin to exponentially back off re-launching new VMIs to replace the failed VMIs. Once a VM's VMI reaches the "running" phase (or the VM is manually stopped) the crashloop tracking will be reset.

## Implementation

Crashloop detection is tracked by a new field on the VM status call `StartFailure` which tracks the number of concurrent start failures as well as tracking when the next start can be retried after a failure.

When a crashloop occurs, users receive feedback that their VM is in a crash loop via the VM.Status.PrintableStatus being set to "CrashLoop"

## Changes in existing behavior

Users can now call `virctl stop my-vm` for VMs with runStrategy Always|RerunOnFailure even when an active VMI is not present. This allows someone to "stop" a VM which is in a crash loop. Without this change in logic, a user wouldn't be able to use virtclt stop during a crash loop because the virt-api subresource endpoint always expects an active VMI in order to stop. 

## Testing

Unit test coverage exists for all new and altered functionality

New functional tests exist to invoke a crashloop, verify exponential backoff occurs, and verify crash loops recover as expected once a VM's vmi eventually hits a running phase.

related to: https://bugzilla.redhat.com/show_bug.cgi?id=1973852

**Release note**:
```release-note
VM CrashLoop detection and Exponential Backoff
```

zcahana reviewed (commented):
Hi @davidvossel, this looks like a very useful addition!
I've included several minor comments inline.

And another general comment:
Currently, the crash loop detection only considers VMIs that never reached the Running phase. However, "Running" VMIs too might end up in some crash loop scenario, in which the guest is launched successfully but quickly fails (boot error, misconfigured liveness probe, ...). Would it make sense to extend it to detect such scenarios too?

davidvossel commented:
> Currently, the crash loop detection only considers VMIs that never reached the Running phase. However, "Running" VMIs too might end up in some crash loop scenario, in which the guest is launched successfully but quickly fails (boot error, misconfigured liveness probe, ...). Would it make sense to extend it to detect such scenarios too?

it's possible we can extend this backoff behavior to failing `liveness` probes. I feel better about the liveness probe situation as it is now because it at least involves some timeout period before the VMI will get rescheduled. That wasn't the case for VMIs that hit some critical error that prevented the qemu process from ever launching successfully (which is represented by Phase: Running)

zcahana commented:
> I feel better about the liveness probe situation as it is now because it at least involves some timeout period before the VMI will get rescheduled

I see your point. I think the main benefit of the CrashLoop detection for Running VMs is the fact that this now gets reported to the user, which otherwise would only see Stopped --> Starting --> Running cycle.

rmohr commented:
Can we also re-evaluate the need for this timeout when the pod hits the running phase? I am not sure we should still keep it. Maybe we should just keep running and wait.

davidvossel commented:
@davidvossel add jitter to launcher timeout

davidvossel commented on cmd/virt-launcher/virt-launcher.go:
```diff
@@ -348,6 +348,7 @@ func main() {
 	qemuAgentUserInterval := pflag.Duration("qemu-agent-user-interval", 10, "Interval in seconds between consecutive qemu agent calls for user command")
 	qemuAgentVersionInterval := pflag.Duration("qemu-agent-version-interval", 300, "Interval in seconds between consecutive qemu agent calls for version command")
 	qemuAgentFSFreezeStatusInterval := pflag.Duration("qemu-fsfreeze-status-interval", 5, "Interval in seconds between consecutive qemu agent calls for fsfreeze status command")
+	simulateCrash := pflag.Bool("simulate-crash", false, "Causes virt-launcher to immediately crash. This is used by functional tests to simulate crash loop scenarios.")
```
yeah, i'm not aware of a reliable method that will immediately fail.

davidvossel commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -748,6 +775,125 @@ func (c *VMController) startVMI(vm *virtv1.VirtualMachine) error {
+	// add randomized seconds to offset multiple failing VMs from one another
+	delaySeconds += rand.Intn(randomRange)
+
+	if delaySeconds > maxDelay {
```
> Otherwise we might end up with multiple failing VMs all stuck at maxDelay


we're adding jitter to the backoff which causes multiple VMs stuck in crash loops not get retried at the same time. by the time max delay is hit, all VMs which are stuck in a crash loop will at least offset from one another due to the randomization leadign up to the max delay

davidvossel commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -748,6 +775,125 @@ func (c *VMController) startVMI(vm *virtv1.VirtualMachine) error {
+			if ts.Phase == virtv1.Running {
+				return true
+			}
+		}
```
+1 fixed

davidvossel commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -748,6 +775,125 @@ func (c *VMController) startVMI(vm *virtv1.VirtualMachine) error {
+	return false
+}
+
+func hasStartFailureBackoffExpired(vm *virtv1.VirtualMachine) int64 {
```
fixed

davidvossel commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -748,6 +775,125 @@ func (c *VMController) startVMI(vm *virtv1.VirtualMachine) error {
+			ConsecutiveFailCount: count,
+		}
+		vm.Status.StartFailure.LastFailedVMIUID = vmi.UID
+		vm.Status.StartFailure.RetryAfterTimestamp = &retryAfter
```
yup, good catch, fixed

davidvossel commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -58,6 +61,8 @@ const (
 	failureDeletingVmiErrFormat           = "Failure attempting to delete VMI: %v"
 )
 
+const defaultMaxCrashLoopBackoffDelay = 300
```
I like working with ints here, i think the issue is i didn't specify that this variable is in seconds.

I renamed this to `defaultMaxCrashLoopBackoffDelaySeconds`

davidvossel commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -748,6 +775,125 @@ func (c *VMController) startVMI(vm *virtv1.VirtualMachine) error {
 }
 
+// Returns in seconds how long to wait before trying to start the VM again.
+func calculateStartBackoffTime(failCount int, maxDelay int) int {
```
since all the calculations are dealing with seconds represented as ints, i'd rather keep it that way.

davidvossel commented on staging/src/kubevirt.io/client-go/api/v1/types.go:
```diff
@@ -1179,6 +1184,8 @@ const (
 	// as well as its associated resources (VirtualMachineInstance, DataVolumes, …).
 	VirtualMachineStatusTerminating VirtualMachinePrintableStatus = "Terminating"
+	// VirtualMachineStatusCrashLoop indicates that the virtual machine is currently in a crash loop waiting to be retried
+	VirtualMachineStatusCrashLoop VirtualMachinePrintableStatus = "CrashLoop"
```
yup, fixed

davidvossel commented:
@zcahana  Thanks for the review. I think I've addressed all your feedback

@rmohr I added jitter to the virt-launcher qemu timeout. The timeout is now a randomized range between 4 and 6 minutes

davidvossel commented:
/retest

zcahana commented:
> @zcahana Thanks for the review. I think I've addressed all your feedback

Thanks David.
/lgtm

davidvossel commented:
/retest

davidvossel commented on pkg/virt-controller/watch/vmi_test.go:
```diff
@@ -1282,6 +1283,10 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
 
 			if vmExists {
 				vmSource.Add(vm)
+				// the controller isn't using informer callbacks for the VM informer
```
the VM controller and VMI controller share the same VM informer. There has to be a VM in the informer cache for the VM controller to create the VMI.

I do need to add a `vmInfomer.HasSynced()` at the vmi controller startup though to prevent the VMI controller from starting until the VM informer is up to date.

zcahana replied:
I see your point, so basically this can only happen at the test since the VM and VMI are fed to the informers "in parallel".
Anyway, looking at the code, you've already added a `vmInformer.HasSynced()` to the VMI controller's `Run()` method, so it's good to go.

davidvossel replied:
> I do need to add a vmInfomer.HasSynced() at the vmi controller startup though to prevent the VMI controller from starting until the VM informer is up to date.

ah, i already did that, never mind

zcahana commented:
/lgtm

davidvossel commented:
/retest

davidvossel commented:
/retest

zcahana commented:
/lgtm

Will need an approver too.

davidvossel commented:
> Will need an approver too.

/cc @rmohr

rmohr commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -578,6 +588,13 @@ func (c *VMController) startStop(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualM
 
+		timeLeft := startFailureBackoffTimeLeft(vm)
+		if timeLeft > 0 {
+			log.Log.Object(vm).Infof("Delaying start of VM %s with 'runStrategy: %s' due to start failure backoff. Waiting %d more seconds before starting.", startingVmMsg, runStrategy, timeLeft)
```
This sounds like something which I would want to see as a warning event.

rmohr commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -724,6 +748,9 @@ func (c *VMController) startVMI(vm *virtv1.VirtualMachine) error {
 	vmi := c.setupVMIFromVM(vm)
+	// add a finalizer to ensure the VM controller has a chance to see
+	// the VMI before it is deleted
+	vmi.Finalizers = append(vmi.Finalizers, virtv1.VirtualMachineControllerFinalizer)
```
I think you mentioned that you want this to avoid issues if the VMI disappears. For me that means that this can still happen (since nothing blocks one from removing finalizers).

It is ok for me to have the finalizer if we have to clean up things. Is that the case? Otherwise I would prefer not to add a finalizer.

rmohr reviewed (commented):
@davidvossel can you explain a little bit in the PR description why a finalizer is now needed?

rmohr commented:
/approve

zcahana commented on pkg/virt-controller/watch/vm_test.go:
```diff
@@ -673,6 +679,214 @@ var _ = Describe("VirtualMachine", func() {
+					// since the tests all run in parallel, it's difficult to
+					// do precise timing. We set the `retryAfter` time but the test
+					// execution may happen seconds later. We use big numbers and
+					// account for some jitter to make sure the calculation falls within
+					// the ballpark of what we expect.
+					parallelTestJitter := expected / 10
+					if (expected - seconds) > parallelTestJitter {
+						Expect(seconds).To(Equal(expected))
+					}
+				}
+
+			},
+
+				table.Entry("no vm start failures",
+					&v1.VirtualMachine{},
+					int64(0)),
+				table.Entry("vm failure waiting 300 seconds",
+					&v1.VirtualMachine{
+						Status: v1.VirtualMachineStatus{
+							StartFailure: &v1.VirtualMachineStartFailure{
+								RetryAfterTimestamp: &metav1.Time{
+									Time: time.Now().Add(300 * time.Second),
+								},
+							},
+						},
+					},
+					int64(300)),
+				table.Entry("vm failure 300 seconds past retry time",
+					&v1.VirtualMachine{
+						Status: v1.VirtualMachineStatus{
+							StartFailure: &v1.VirtualMachineStartFailure{
+								RetryAfterTimestamp: &metav1.Time{
+									Time: time.Now().Add(-300 * time.Second),
+								},
+							},
+						},
+					},
+					int64(0)),
+			)
+		})
```
I realize this is merged for quite some time, but resurrecting this following a unit tests failure I saw on the CI ([this](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6190/pull-kubevirt-goveralls/1423671111240912896)).
Looking closely at this test, I reckon that:
- In table entries 1 and 3, `expected` is always 0, so nothing is actually being asserted (`if expected > 0 { ... }`).
- In entry 2, if test execution happens >30 seconds after the evaluation of the entry, it's bound to fail, and otherwise silently pass without any assertion being made.

Actually, I'm not entirely sure what the purpose of this specific test. Also, given the logic of `calculateStartBackoffTime()` is quite simple and is anyway covered by the larger testcases in the `Context("crashloop backoff tests")`, do you think we could just remove it?

@davidvossel WDYT?

zcahana replied:
#6203
